DataFormats/Provenance/interface/IndexIntoFile.h

```diff
@@ -247,15 +267,15 @@ namespace edm {
     /// This enum is used to specify the order of iteration.
     /// In firstAppearanceOrder there are 3 sort criteria, in order of precedence these are:
     ///
-    ///   1. firstAppearance of the ProcessHistoryID and run number in the file
+    ///   1. firstAppearance of the ProcessHistoryID and run number in the output module
```

Comment:
Before concurrent lumis and concurrent runs, the order was the order in the file. Specifically, the sort order used by IndexIntoFile was the same as the order in the Run TTree of the file. The Events TTree and Lumi TTree would have the same order except that some Runs might not contain any Events or any Lumis. In the process creating an IndexIntoFile, the order in the input file and the output file would be the same.

In a process with concurrent runs and concurrent lumis, this is no longer true. Everything gets scrambled. The order in the Run, Lumi, and Events TTree's can be different in the same file. The order in the input and output files can be different. It is not even well defined what "file order" means anymore. On top of that, the IndexIntoFile sort order criteria can be different than all of those other orderings. What really determines the order is the timing of the calls to writeEvent, writeLumi, and writeRun. Whichever of those three function calls to the output module occurs first, establishes which Run is next. It is related to the file order, close to the file order, but not exactly the same as the order in any TTree in the input or output file.

I could force the ordering, but that would add complication, waits, caching...

Reply:
> Before concurrent lumis and concurrent runs, the order was the order in the file

I assume you mean the order in the _input file being read by the Source_.

Here, the original documentation
```
1. firstAppearance of the ProcessHistoryID and run number in the file
```
I interpret as _file_ to be the file containing THIS IndexIntoFile, not the _input file read from the Source in the job that created the file now holding this IndexIntoFile_.

I'd prefer the documentation in IndexIntoFile to say how it relates to other meta data in the TFile which is holding this IndexIntoFile rather than relating to the mechanics of the job which created that TFile.

Reply:
For runs that have at least one event, then the order will be the same as in the Events TTree in the file. If that is not true though, then I think it is not possible to determine the order from the file. I can't think of a way.

I could say "firstAppearance of the ProcessHistoryID and run number in the Events TTree of the file (in cases where there are no Events in the run it depends on the order of the call to writeRun or writeLumi in the preceding step where the IndexIntoFile was created)" Is that OK?

Reply:
OK by me.